Filter integer header names during SAPI discovery

weierophinney opened the pull request:
|    Q          |   A
|-------------- | ------
| Documentation | yes
| Bugfix        | yes
| BC Break      | yes
| New Feature   | no
| RFC           | no
| QA            | no

### Description

Integer strings have the nasty habit of being cast to actual integers by PHP, making them problematic for usage in an associative array, despite being valid per the RFC 7230 ABNF.

Additionally, having them pass through `marshal_headers_from_sapi()` means that once `ServerRequest` gets them and tries to use them, `HeaderSecurity::assertValidName()` will raise an exception for integers, which could lead to unexpected server errors.

This patch chooses to filter such header names out entirely.
It also documents the change, and ways to address it if you previously depended on integer header field names.

Fixes #11

weierophinney committed:
feature: filter out integer keys when marshaling headers from SAPI

Integer strings have the nasty habit of being cast to actual integers by PHP, making them problematic for usage in an associative array, despite being valid per the RFC 7230 ABNF.

Additionally, having them pass through `marshal_headers_from_sapi()` means that once `ServerRequest` gets them and tries to use them, `HeaderSecurity::assertValidName()` will raise an exception for integers, which could lead to unexpected server errors.

This patch chooses to filter such header names out entirely.
Doing so prevents those server errors.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
docs: document change to filter integer header names

Creates a "security features" document that brings in the former v2 "forward migration" document around filtering x-forwarded-* headers, and also adds narrative around filtering integer headers.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
docs: adds section to migration doc to cover marshalHeadersFromSapi change

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

Xerkus commented on src/functions/marshal_headers_from_sapi.php:
```diff
@@ -63,5 +66,6 @@ function marshalHeadersFromSapi(array $server): array
     }
 
-    return $headers;
+    // Filter out integer keys
+    return array_filter($headers, fn(string|int $key): bool => is_string($key), ARRAY_FILTER_USE_KEY);
```
Please document inline where the integer keys come from.

Xerkus reviewed (commented):
`ServerRequestFactory::fromGlobals()` needs test to ensure it filters out integer-like headers since `marshalHeadersFromSapi()` is an implementation detail.

Xerkus commented on docs/book/v3/migration.md:
```diff
@@ -24,6 +24,12 @@ For consumers, usage should be completely backwards compatible, however.
 
+### marshalHeadersFromSapi
+
+The function `Laminas\Diactoros\marshalHeadersFromSapi()`, which is consumed by `Laminas\Diactoros\ServerRequestFactory::fromGlobals()`, was modified such that it will no filter out header field names that evaluate to integers.
```
typo
```suggestion
The function `Laminas\Diactoros\marshalHeadersFromSapi()`, which is consumed by `Laminas\Diactoros\ServerRequestFactory::fromGlobals()`, was modified such that it will now filter out header field names that evaluate to integers.
```

Xerkus commented on docs/book/v3/security.md:
````diff
@@ -0,0 +1,39 @@
+1234: header value
+```
+
+The PSR-7, `Psr\Http\MessageInterface::getHeaders()` method requires implementations to return an associative array, where the key is the header field name.
+This triggers an interesting quirk in PHP: when adding a value to an array using a string that consists of an integer value, PHP will convert this value to an integer.
````
Should this link to original php bug report?
https://bugs.php.net/bug.php?id=80309

Xerkus commented on docs/book/v3/security.md:
```diff
@@ -0,0 +1,39 @@
+- First, it means that consumers cannot depend on the header field name returned being a string.
+- Second, our own validation of header field name will fail, as it will not see a string.
+
+Normally, this will not present an issue, as the way to add headers to a message is via the `MessageInterface::withHeader()` and `MessageInterface::withAddedHeader()` methods, which both require a `string` name argument.
```
note: `string` on name argument does not prevent header from being added as `'1234'`. It will definitely present an issue, just not immediately as on PSR-7 message creation where it had chance to convert to int before validation.

weierophinney replied:
Correct. 

In fact... in that case, `getHeaders()` will still return the key as well. Which means there's no guarantee on the consumer end that the array from that has only string keys.

I'll add a note to that effect.

Xerkus reviewed (changes requested):
A typo, otherwise good.

weierophinney committed:
docs: document why we filter integer keys from the $headers array

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: add integer keys to header lists while testing

Adds integer and string-integer keys from the `marshalHeadersFromSapi()` test case to the ServerRequestFactory test cases to demonstrate they get filtered out.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
Update docs/book/v3/migration.md

Co-authored-by: Aleksei Khudiakov <aleksey@xerkus.pro>
Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
docs: link to PHP bug 80309

Links to PHP bug 80309 in discussion of the string integer value to integer conversion.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
docs: add a note about integer keys added via other means

When a header field is added with `withHeader()` or `withHeaderLine()`, these will accept strings that have digits only, which can also lead to the issues presented elsewhere.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
docs: remove extra linebreak

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

Xerkus commented on src/functions/marshal_headers_from_sapi.php:
```diff
@@ -63,5 +66,9 @@ function marshalHeadersFromSapi(array $server): array
 
-    return $headers;
+    // Filter out integer keys.
+    // These can occor if the translated header name is a string integer.
+    // PHP will cast those to integers when assigned to an array.
+    // This filters them out.
```
Typo

```suggestion
    // These can occur if the translated header name is a string integer.
    // PHP will cast those to integers when assigned to an array.
    // This filters them out.
```

weierophinney committed:
Update src/functions/marshal_headers_from_sapi.php

Co-authored-by: Aleksei Khudiakov <aleksey@xerkus.pro>
Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>
